### PR TITLE
Feature: Clean architecture entities, controllers and services

### PR DESCRIPTION
**Pull request by yeahlowflicker**
引入 Clean Architecture 概念，包含 Entity, Controller 和 Service：
- Entity: 資料模型，以物件方式運作
- Controller: 後端控制器，負責與 Supabase 溝通
- Service: 負責資料轉換、檢查

每個 Entity 內會多含一個 DBXXXX 的 type，如 User 有 DBUser，後者是 Supabase 產生的 type，會透過 Service 轉成正常的 entity type。

`ErrorHandler` 是通用的錯誤處理器，供後續作 error display、狀態更新使用。

其餘細節、用法請參閱各 function 註解。

(2024/12/14)

## Summary by CodeRabbit

- **New Features**
	- Introduced a universal error handler for improved error management.
	- Added functionality to retrieve and find events and users from the database.
	- Implemented user identity conversion for better user representation.

- **Bug Fixes**
	- Enhanced error handling during data retrieval operations.

- **Documentation**
	- Added documentation comments for new classes and methods to improve clarity.

### CODE REVIEW

**Review by daviddaytw — COMMENTED** [review]
@yeahlowflicker Fantastic work!

This PR is very important since it will solidify future code styles.

I'm wondering if Entities and Services are really necessary since they seem redundant. What are the benefits?
Also, what does `features` mean actually? Is there a better vocabulary for it?

**Comment by yeahlowflicker** [comment]
@daviddaytw Thank you David.

I introduced the entities because of the following 3 reasons:

1. Ensure properties are not null by setting default values within the entities. This may be useful when fetching partial columns.

2. Facilitate data transfer within the frontend by using unified types.

3. Provide utility functions directly in the entity objects, like data validation methods and text mapping.

Regarding the folder name, what do you think about the term `modules`?

These are based on my limited understanding of the Supabase types. Please feel free to correct me if there's anything wrong, thanks.

**Comment by daviddaytw** [comment]
I see the reasons. I'm okay with that if an internal consensus is made.
In my understanding, the folder will be everything related to the database or backend. Can we use either one as the name so it is self-explaining?

**Review by 1989ONCE — APPROVED** [review]

**Comment by 1989ONCE** [comment]
@daviddaytw Thanks for the advice! I think `backend` could be a solid choice 🤩
@yeahlowflicker Does backend work for you, or do you have other ideas? Would love to hear your thoughts!

**Comment by yeahlowflicker** [comment]
@daviddaytw @1989ONCE
`backend` seems logical to me, thanks.

File: src/backend/event/Services/EventService.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import Event, { DBEvent } from "../Entities/Event"
+
+export default class EventService {
+    
+    public static parseEvent(record: DBEvent) : Event {
```
[the remark this answers is not in the document]
Comment (daviddaytw) [reply src/backend/event/Services/EventService.ts:5]: @yeahlowflicker There is a better way to do this, but it's up to you.

```typescript
const EventService = {
    parseEvent: (record: DBEvent) => {
    }
}
export default EventService
```

**Review by daviddaytw — APPROVED** [review]

**Review by 1989ONCE — APPROVED** [review]

**Review by daviddaytw — APPROVED** [review]
LGTM!

**Pull request merged**